scripts/firedrake-install

```diff
@@ -1030,6 +1031,31 @@ def build_and_install_petsc():
                 check_call(["rm", "-f", deletefile])
 
 
+def build_and_install_petsc4py():
+    with environment(**blas):
+        if os.access(get_petsc4py_dir(), os.W_OK):
+            pipargs = ["-vvv", "--ignore-installed", get_petsc4py_dir()]
+            check_output(pipinstall + pipargs)
+        else:
+            log.info("Could not perform in-place install of petsc4py in current location")
+            log.info("Copying petsc4py to virtualenv and trying again...")
+            petsc4py_copy = os.path.join(os.environ['VIRTUAL_ENV'], 'src', 'petsc4py')
+            if os.path.exists(petsc4py_copy):
+                petsc4py_backup = os.path.join(os.environ['VIRTUAL_ENV'], 'src', 'petsc4py.backup')
+                if os.path.exists(petsc4py_backup):
+                    raise OSError('`{}` exists remove or rename this file/directory and try running this command again'.format(petsc4py_backup))
+                shutil.move(petsc4py_copy, petsc4py_backup)
```

Comment:
I would much rather do the following:

1. Check if source dir is writeable (ok, just go as normal)
2. If not, copy somewhere (really not sure why you need to do the dance with mode changes [if I copy something just at the commandline, it gets the correct permissions], can you explain?)
3. Build and install
4. If the build is successful, remove the copied directory
5. Otherwise leave it around for debugging [but put a README file in it that indicates that this is not the actual place that the source code lives]

WDYT?

Reply:
I think this is a good idea, I think I would modify the procedure slightly based on trying to install from PETSc installed by a package manager/system administrator. It turns out checking to see if the source is writeable isn't sufficient for the install to complete. For instance if build artefacts are present in petsc4py left over from any other user building petsc4py in the same writeable directory. This seems especially relevant given the issues that prompted this PR.

My modification would be:
1. Always copy the source dir to the venv (rather than trying to program every possible exception)
Then 3, 4 and 5 as above.

I'm not sure there is an automatic solution to update petsc4py if the system PETSc changes, users will have to use the `--rebuild` flag